presto-record-decoder/src/main/java/io/prestosql/decoder/json/AbstractDateTimeJsonValueProvider.java

```diff
@@ -70,10 +70,10 @@ public final long getLong()
         if (type.equals(TIME)) {
             return millis * PICOSECONDS_PER_MILLISECOND;
         }
-        if (type.equals(TIMESTAMP)) {
+        if (type instanceof TimestampType) {
```

Comment:
I don't think this is right.  If the precision is great than 3, you need a different encoding.

Reply:
Kafka [table definition files](https://prestosql.io/docs/current/connector/kafka.html#table-definition-files) only allow specifying `TIMESTAMP` without precision for field mappings.  So at this point, Kafka table columns can only have millisecond precision for timestamps.  This is a valid concern though for the code that handles writes, though, since the source data could potentially have columns with higher precision.

Reply:
`presto-record-decoder` is a module that can be used by other connectors aside from Kafka. If this code declares that it supports any precision for `timestamp`, it should produce the correct value:
* for p <= 3, a long representing milliseconds from the epoch
* for 3 < p <= 6, a long representing microseconds from the epoch
* for p > 6, an instance of LongTimestamp

Reply:
I changed it back to an equality check with the timestamp singleton. Same goes for tswtz. Supporting variable precision timestamps in the record decoder is a change for a different pr.